src/modules/rlm_json/json.c

```diff
@@ -252,3 +252,105 @@ void fr_json_version_print(void)
 #endif
 }
 
+static json_object *add_vp_to_json_array(TALLOC_CTX *ctx, json_object *arr, VALUE_PAIR *vp) {
+	json_object *to_add = NULL;
+	char *tmp;
+	size_t len, outlen;
+
+	switch (vp->da->type) {
+		case PW_TYPE_INTEGER:
+			to_add = json_object_new_int(vp->vp_integer);
+			break;
+		case PW_TYPE_SHORT:
+			to_add = json_object_new_int(vp->vp_short);
+			break;
+		case PW_TYPE_BYTE:
+			to_add = json_object_new_int(vp->vp_byte);
+			break;
+		case PW_TYPE_SIGNED:
+			to_add = json_object_new_int(vp->vp_signed);
+			break;
+		case PW_TYPE_INTEGER64:
+			to_add = json_object_new_int64(vp->vp_integer64);
+			break;
+		case PW_TYPE_BOOLEAN:
+			to_add = json_object_new_boolean(vp->vp_byte);
+			break;
+		default:
+			len = vp->data.length+1;
+			tmp = NULL;
+			do {
+				if (tmp) talloc_free(tmp);
+				tmp = talloc_array(ctx, char, len);
+				if (!tmp) return NULL;
+				outlen = fr_pair_value_snprint(tmp, len, vp, 0);
+				len <<= 1;
+			} while (outlen >= len);
+			to_add = json_object_new_string(tmp);
+			talloc_free(tmp);
+			break;
+	}
+
+	if (to_add)
+		json_object_array_add(arr, to_add);
+
+	return to_add;
+}
+
+const char *fr_json_vps_to_json(TALLOC_CTX *ctx, VALUE_PAIR **vps, const char *prefix) {
+	vp_cursor_t cursor;
+	struct json_object *obj, *vp_object, *values, *type_name;
+	VALUE_PAIR *vp, *vp_next;
+	char *name_with_prefix;
+	const char *res = NULL;
+
+	fr_pair_list_sort(vps, fr_pair_cmp_by_da_tag);
+	vp = fr_cursor_init(&cursor, vps);
+
+	obj = json_object_new_object();
+	if (!obj) return NULL;
+	while (vp) {
+		VERIFY_VP(vp);
+
+		vp_object = json_object_new_object();
+		if (!vp_object) goto fr_json_vps_to_json_failure;
+
+		if (prefix) {
+			name_with_prefix = talloc_zero_array(ctx, char, strlen(prefix) + strlen(vp->da->name) + 2);
+			if (!name_with_prefix) {
+				json_object_put(vp_object);
+				goto fr_json_vps_to_json_failure;
+			}
+			sprintf(name_with_prefix, "%s:%s", prefix, vp->da->name);
```

Comment:
Also these only fail on memory allocation errors, so wrapping with ``MEM()`` is sufficient.